models/shift_net/InnerShiftTripleFunction.py

```diff
@@ -30,34 +29,53 @@ def forward(ctx, input, shift_sz, stride, triple_w, flag, show_flow):
         ctx.flag = ctx.flag.to(input).long()
 
         # None batch version
-        Nonparm = Modified_NonparametricShift()
+        bNonparm = Batch_NonShift()
         ctx.shift_offsets = []
 
-        for idx in range(ctx.bz):
-            flag_cur = ctx.flag[idx]
-            latter = latter_all.narrow(0, idx, 1) ### encoder feature
-            former = former_all.narrow(0, idx, 1) ### decoder feature
+        # batch version
+        cosine, latter_windows, i_2, i_3, i_1 = bNonparm.cosine_similarity(former_all.clone(), latter_all.clone(), 1, stride, flag)
 
-            #GET COSINE, RESHAPED LATTER AND ITS INDEXES
-            cosine, latter_windows, i_2, i_3, i_1, i_4 = Nonparm.cosine_similarity(former.clone().squeeze(), latter.clone().squeeze(), 1, stride, flag_cur)
+        _, indexes = torch.max(cosine, dim=2)
 
-            ## GET INDEXES THAT MAXIMIZE COSINE SIMILARITY
-            _, indexes = torch.max(cosine, dim=1)
+        mask_indexes = (flag==1).nonzero()[:, 1].view(ctx.bz, -1)
 
-            # SET  TRANSITION MATRIX
-            mask_indexes = (flag_cur == 1).nonzero()
-            non_mask_indexes = (flag_cur == 0).nonzero()[indexes]
-            ctx.ind_lst[idx][mask_indexes, non_mask_indexes] = 1
+        # These three lines of code work similarly with `non_mask_indexes_tmp[indexes]`
+        # However, pytorch advance indexing dose not support multi-dimensional indexing.
+        non_mask_indexes_tmp = (flag==0).nonzero()[:,1].view(ctx.bz, -1)
+        idx_tmp = indexes + torch.arange(indexes.size(0)).view(-1,1).type_as(indexes) * non_mask_indexes_tmp.size(1)
+        non_mask_indexes = non_mask_indexes_tmp.view(-1)[idx_tmp]
 
-            # GET FINAL SHIFT FEATURE
-            shift_masked_all[idx] = Nonparm._paste(latter_windows, ctx.ind_lst[idx], i_2, i_3, i_1, i_4)
+        # Update: batch op if someone think up an approach
+        for i in range(ctx.bz):
```

Comment:
@tchaton It is possible to remove `for loop` here ? It should be the final snippet that isn't implemented in batch version.

Reply:
@Zhaoyi-Yan , I think it is possible to vectorize everything there.

Reply:
@tchaton I have ask in the pytorchDiscuss and get a nice solution. I have updated it.